azurerm/internal/services/cognitive/cognitive_account_customer_managed_key_resource.go

```diff
@@ -0,0 +1,253 @@
+	}
+
+	stateConf := &pluginsdk.StateChangeConf{
+		Pending:    []string{"Accepted"},
+		Target:     []string{"Succeeded"},
+		Refresh:    cognitiveAccountStateRefreshFunc(ctx, client, *id),
+		MinTimeout: 15 * time.Second,
+		Timeout:    d.Timeout(pluginsdk.TimeoutCreate),
+	}
+
+	if _, err = stateConf.WaitForStateContext(ctx); err != nil {
+		return fmt.Errorf("waiting for update of %s: %+v", *id, err)
+	}
+	d.SetId(id.ID())
+
+	return resourceCognitiveAccountCustomerManagedKeyRead(d, meta)
+}
+
+func resourceCognitiveAccountCustomerManagedKeyRead(d *pluginsdk.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Cognitive.AccountsClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.AccountID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.Name)
+	if err != nil {
+		return fmt.Errorf("retrieving %s: %+v", *id, err)
+	}
+	if resp.Properties == nil && resp.Properties.Encryption == nil {
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("cognitive_account_id", id.ID())
+	d.Set("key_source", resp.Properties.Encryption.KeySource)
+	if props := resp.Properties.Encryption.KeyVaultProperties; props != nil {
+		var keyName string
+		if props.KeyName != nil {
+			keyName = *props.KeyName
+		}
+
+		var keyVaultUri string
+		if props.KeyVaultURI != nil {
+			keyVaultUri = *props.KeyVaultURI
+		}
+
+		var keyVersion string
+		if props.KeyVersion != nil {
+			keyVersion = *props.KeyVersion
+		}
+		keyVaultKeyId, err := keyVaultParse.NewNestedItemID(keyVaultUri, "keys", keyName, keyVersion)
+		if err != nil {
+			return fmt.Errorf("parsing `key_vault_key_id`: %+v", err)
+		}
+		d.Set("key_vault_key_id", keyVaultKeyId.ID())
+		if props.IdentityClientID != nil {
+			d.Set("identity_client_id", *props.IdentityClientID)
+		}
+	} else {
+		d.Set("key_vault_key_id", nil)
+		d.Set("identity_client_id", nil)
+	}
+
+	return nil
+}
+
+func resourceCognitiveAccountCustomerManagedKeyDelete(d *pluginsdk.ResourceData, meta interface{}) error {
+	accountsClient := meta.(*clients.Client).Cognitive.AccountsClient
+	deletedAccountsClient := meta.(*clients.Client).Cognitive.DeletedAccountsClient
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.AccountID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	locks.ByName(id.Name, "azurerm_cognitive_account")
+	defer locks.UnlockByName(id.Name, "azurerm_cognitive_account")
+
+	account, err := accountsClient.Get(ctx, id.ResourceGroup, id.Name)
+	if err != nil {
+		return fmt.Errorf("retrieving %s: %+v", *id, err)
+	}
+
+	// Since this isn't a real object and it cannot be disabled once Customer Managed Key at rest has been enabled
+	// And it must keep at least one key once Customer Managed Key is enabled
+	// So for the delete operation, it has to recreate the Cognitive Account with disabled Customer Managed Key
```

Comment:
We can't just delete and recreate the associated Cognitive account?

With these limitations, this resource likely isn't viable. Perhaps it would be better as a block on the `azurerm_cognitive_account` resource?

Reply:
Hi @manicminer , we can't add it to `azurerm_cognitive_account`. Because we have a key access policy depends on `azurerm_cognitive_account#identity`, and a key vault key depends this policy, then cmk depends on the key vault key.
This means if we add it to `azurerm_cognitive_account`, it will cause circular dependency.

So it has to be a separate resource like other existing cmk resource.

refs: https://github.com/terraform-providers/terraform-provider-azurerm/pull/12159

Reply:
Hi @manicminer , we can't add it to `azurerm_cognitive_account`. Because we have a key access policy depends on `azurerm_cognitive_account#identity`, and a key vault key depends this policy, then cmk depends on the key vault key.
This means if we add it to `azurerm_cognitive_account`, it will cause circular dependency. 

So it has to be a separate resource like other existing cmk resource.

refs: https://github.com/terraform-providers/terraform-provider-azurerm/pull/12159